### DISCUSSION TITLE
pin devDeps for AwsCdkTypeScriptApp

### DISCUSSION THREAD

**Discussion by jlosito** [projen/projen#3485] [Q&A] [answered]
I have a requirement where all dependencies have to be pinned to a specific version. This includes devDependencies. Using v0.80.10 of projen I'm trying to pin devDepdencies as such, but I still see a range being included in the output.

```
# .projenrc.ts
import { awscdk } from "projen";
import { TrailingComma } from "projen/lib/javascript";
onst project = new awscdk.AwsCdkTypeScriptApp({
	cdkVersion: "2.111.0",
	cdkVersionPinning: true,
	defaultReleaseBranch: "main",
	name: "my-cdk-app-project",
	projenrcTs: true,

	depsUpgrade: false,
	deps: ["cdk-nag@2.28.65"],
	devDeps: ["esbuild@0.20.1"],

	gitOptions: {},

	github: false,

	prettier: true,
	prettierOptions: {
		settings: {
			tabWidth: 4,
			useTabs: true,
			trailingComma: TrailingComma.ALL,
			printWidth: 140,
		},
	},
	eslintOptions: {
		prettier: true,
		dirs: ["src", "src/**", "test", "test/**"],
	},
});

project.synth();
```

Here's what I see in the output. While dependencies do get pinned, dev ones don't.
```
  "devDependencies": {
    ...
    "esbuild": "^0.20.1",
    ...
    },
    "dependencies": {
    "aws-cdk-lib": "2.111.0",
    "cdk-nag": "2.28.65",
    "constructs": "^10.0.5"
  },
```

**Comment by mrgrain** [comment]
So `constructs` and `esbuild`? Can you check `.projen/deps.json` for me and post the relevant entries? Note that there might be multiple ones for each dependency.

**Comment by jlosito** [comment]
```
{
  "dependencies": [
    {
      "name": "@types/jest",
      "type": "build"
    },
    {
      "name": "@types/node",
      "version": "^18",
      "type": "build"
    },
    {
      "name": "@typescript-eslint/eslint-plugin",
      "version": "^6",
      "type": "build"
    },
    {
      "name": "@typescript-eslint/parser",
      "version": "^6",
      "type": "build"
    },
    {
      "name": "aws-cdk",
      "version": "2.111.0",
      "type": "build"
    },
    {
      "name": "esbuild",
      "type": "build"
    },
    {
      "name": "eslint-config-prettier",
      "type": "build"
    },
    {
      "name": "eslint-import-resolver-typescript",
      "type": "build"
    },
    {
      "name": "eslint-plugin-import",
      "type": "build"
    },
    {
      "name": "eslint-plugin-prettier",
      "type": "build"
    },
    {
      "name": "eslint",
      "version": "^8",
      "type": "build"
    },
    {
      "name": "jest",
      "type": "build"
    },
    {
      "name": "jest-junit",
      "version": "^15",
      "type": "build"
    },
    {
      "name": "prettier",
      "type": "build"
    },
    {
      "name": "projen",
      "type": "build"
    },
    {
      "name": "ts-jest",
      "type": "build"
    },
    {
      "name": "ts-node",
      "type": "build"
    },
    {
      "name": "typescript",
      "type": "build"
    },
    {
      "name": "aws-cdk-lib",
      "version": "2.111.0",
      "type": "runtime"
    },
    {
      "name": "cdk-nag",
      "version": "2.28.65",
      "type": "runtime"
    },
    {
      "name": "constructs",
      "version": "^10.0.5",
      "type": "runtime"
    }
  ],
  "//": "~~ Generated by projen. To modify, edit .projenrc.ts and run \"npx projen\"."
}

```

**Comment by jlosito** [comment]
@mrgrain The same goes for `constructs`. Right now there's only the ability to specify a minimum version for constructs via `constructsVersions` but I'm unable to pin to a specific version like the `cdkVersionPinning` prop.

**Comment by mrgrain** [comment, marked as answer]
tl;dr if you add this after the project creation it works for me:

```ts
project.addDeps('constructs@10.0.5');
project.addDevDeps('esbuild@0.20.1');
```

---

@jlosito Apologies for the late reply. This is effectively an ordering problem and the last dependency definition wins.

1. Your code adds `esbuild@0.20.1`
2. A Component later overwrites this to `esbuild` without constraint which than gets turned into the caret version.

In this case you could have actually just gone into `package.json` and remove the caret and it should stay like that as well.

`constructs` is different because the project type strongly believes you should not pin this dependency (other than via lockfile).

**Comment by jlosito** [comment]
@mrgrain thank you. That totally works.

```
# .projenrc.ts
import { awscdk } from "projen";
import { TrailingComma } from "projen/lib/javascript";
onst project = new awscdk.AwsCdkTypeScriptApp({
	cdkVersion: "2.111.0",
	cdkVersionPinning: true,
	defaultReleaseBranch: "main",
	name: "my-cdk-app-project",
	projenrcTs: true,

	depsUpgrade: false,
	deps: ["cdk-nag@2.28.65"],

	gitOptions: {},

	github: false,

	prettier: true,
	prettierOptions: {
		settings: {
			tabWidth: 4,
			useTabs: true,
			trailingComma: TrailingComma.ALL,
			printWidth: 140,
		},
	},
	eslintOptions: {
		prettier: true,
		dirs: ["src", "src/**", "test", "test/**"],
	},
});

project.addDevDeps("esbuild@0.20.1", "eslint@8.57.0");

project.synth();
```

**Reply by mrgrain** [reply]
Converted this to a discussion for now. I feel there is _some_ feature improvement trying to get out of this, but I'm not quite sure yet what that would look like. 🤔

**Reply by jlosito** [reply]
Given how you described it, it makes more sense to me now why this behavior was happening. As of now, I can't see anything that should change.

**Reply by mrgrain** [reply]
To me it's something along the lines of this: The component that adds `esbuild` clearly doesn't care about the version, so when it finds that there already is a dependency on `esbuild` that is more specific, why does it overwrite the more specific and not conflicting declaration. Same for constructs.
